src/datasets/arrow_dataset.py

```diff
@@ -577,17 +577,22 @@ def save_to_disk(self, dataset_path: str, fs=None):
 
         # Get json serializable state
         state = {
-            key: self.__dict__[key] if key != "_split" else str(self.__dict__[key])
+            key: self.__dict__[key]
             for key in [
                 "_fingerprint",
                 "_format_columns",
                 "_format_kwargs",
                 "_format_type",
                 "_indexes",
                 "_output_all_columns",
-                "_split",
             ]
         }
+
+        split = self.__dict__["_split"]
+        state["_split"] = {"spec": str(split) if split is not None else split}
+        if isinstance(split, (NamedSplit, ReadInstruction)):
```

Comment:
According to the typing, `split` should be either None or a NamedSplit, not a ReadInstruction.
Should we fix the typing as well ?

Another option I have in mind is to always use NamedSplit for the `_split` attribute. I think we would just need to use `to_spec` at this line:

https://github.com/huggingface/datasets/blob/20ffe98d1da381f28464684a62c8ed3a499e10aa/src/datasets/arrow_reader.py#L238-L238

`original_instruction` could be a string, a NamedSplit or a ReadInstruction (damn this function lacks typing -_-)

I have a small preference for this last option because that simplify things and make the serialization a simple string instead of `{"spec": ..., "type": ...}`. Currently this is a string that is used, so that we could avoid breaking backward compatibility this way.

Let me know what you think !

Reply:
Yes, I agree this "spec + type" thing is not ideal and that the `_split` attribute should either be None or a NamedSplit to get a simpler solution.